// SPDX-FileCopyrightText: Copyright (c) 2024 Intel Corporation
//
// SPDX-License-Identifier: BSD-3-Clause

syntax = "proto3";

package sdk;

import "type.proto";
import "conn-config.proto";

service SDKAPI {
  rpc RegisterAndStreamEvents (RegisterRequest) returns (stream Event);

  rpc CreateConnection (CreateConnectionRequest) returns (CreateConnectionResponse);
  rpc ActivateConnection (ActivateConnectionRequest) returns (ActivateConnectionResponse);
  rpc DeleteConnection (DeleteConnectionRequest) returns (DeleteConnectionResponse);
}

message RegisterRequest {
  repeated type.Tag tags = 1;
}

message ClientRegisteredEvent {
  string client_id = 1;
}

message ConnectionUnlinkRequestedEvent {
  string conn_id = 1;
}

message LoggerConfigChangedEvent {
  repeated string filters = 1;
}

message Event {
  oneof event {
    ClientRegisteredEvent client_registered              = 1;
    ConnectionUnlinkRequestedEvent conn_unlink_requested = 2;

    LoggerConfigChangedEvent logger_config_changed       = 100;
  }
}

message CreateConnectionRequest {
  string client_id        = 1;
  bytes mcm_conn_param    = 2;
  ConnectionConfig config = 3;
  string name             = 4;
}

message CreateConnectionResponse {
  string conn_id         = 1;
  bytes memif_conn_param = 2;
}

message ActivateConnectionRequest {
  string client_id = 1;
  string conn_id   = 2;
}

message ActivateConnectionResponse {
  bool linked = 1;
}

message DeleteConnectionRequest {
  string client_id = 1;
  string conn_id   = 2;
}

message DeleteConnectionResponse {
}
